### DISCUSSION TITLE
declare module 'vue' breaks types for auto imported components

### DISCUSSION THREAD

**Discussion by J-Michalek** [vuejs/pinia#2729] [Help and Questions] [answered]
### Reproduction

N/A

### Steps to reproduce the bug

1) Declare global component types under vue module:
```ts
declare module 'vue' {
  export interface GlobalComponents {
    ...
  }
}
```

2) Install pinia 2.2.0
3) Check for TS errors inside of Vue components that make us of the global components

### Expected behavior

It should not break the types for global components.

### Actual behavior

Breaks the types for global components

### Additional information

I am pretty sure this is the change that broke this in our project:
https://github.com/vuejs/pinia/commit/8a6ce86db83b6315c067c8a98c898b3c74efe62e#diff-094806c433011456914201ea662b339b53b9b1236f3226c1da3fd807106ad8a8R43

We use https://github.com/unplugin/unplugin-vue-components and the types it generates seemingly conflict with the ones Pinia declares.

**Comment by posva** [comment]
In the ecosystem we were using `@vue/runtime-core` but the recommended way was to use `vue` and it took some time to adapt. At some point in time, upgrading TS versions and other dependencies **required** this change. In my testing, this change didn't break other libraries using `@vue/runtime-core` **when using up-to-date dependencies and dev dependencies**. You might need to update TS

**Comment by mxp131011** [comment]
I've been in a similar situation

**Comment by J-Michalek** [comment]
We are running on the latest TS version (5.5.4) but the issue is there after i update Pinia.

The only deps that are not up to date seem unrelated to this issue. I have listed them below.

dev deps:
- eslint
- eslint-plugin-chai-friendly

deps:
- vue-final-modal
- vuex

**Comment by xmatthias** [comment]
I think the key here is the use in combination with unplugin-vue-components or unplugin-auto-import.

the generated component.d.ts file contains something around the following, then:

``` typescript
...
export {}

/* prettier-ignore */
declare module 'vue' {
  export interface GlobalComponents {
    MyComponent: typeof import('./components/MyComponent.vue')['default']
    MyOtherComponent: typeof import('./components/MyOtherComponent.vue')['default']
...
}
```

In my case - the update in the dependabot group broke all type-checks.
Reverting (and hard-pinning) pinia fixes this behavior again - with this, pinia is now the only dependency not uptodate.

**edit** while not a good, minimal reproduction (it's just my project breaking) - this PR showcases that the update of pinia breaks check-types (Ci is passing on main):
https://github.com/freqtrade/frequi/pull/2004 

Also noteworthy: the type-check failures are not due to pinia usage, but due to random, other, auto-imported utility functions. (which in hindsight, would make it more reasonable that it's acutally conflicting with `unplugin-auto-import`).

**Comment by posva** [comment]
Probably unrelated, but in your project, you should try out upgrading the moduleResolution to Bundler: https://github.com/freqtrade/frequi/blob/main/tsconfig.json#L8

I tested locally by adding unplugin-vue-components to https://github.com/posva/demo-data-loaders-artwork-gallery, and it works, so I think you have something in your TS config breaking it. I'm moving this to a question so anybody can try to figure it out and propose and answer

**Reply by xmatthias** [reply]
I'm not sure why this is a question to you (not at this point, anyway).
The label says "needs repro" - but people are NOT given the time to provide a reproduction (i do not consider a dependabot PR to a random repository as proper repro).

It's pretty clearly the pinia update breaking this (the only thing changed in the dependabot PR) - so it's a bug in pinia in my eyes - one that the update introduced.

I also think it's not unplugin-vue-components - but unplugin-auto-import which is causing the problematic conflicts - as these are the imported utils that no longer show as "available" (at least in my case).

**Reply by xmatthias** [reply]
@posva as for reproduction:
i've used your artwork-gallery example and added a new branch in my fork: https://github.com/xmatthias/demo-data-loaders-artwork-gallery/tree/feat/autoimport

For demo purposes, i also added a completely pointless "something" utility function designed to show the problem.

When this function is used in a setup / script setup function - it works fine (no type error).
If it's used inside of a template - then it'll type-error.

with ` "pinia": "2.1.7",` - this is not happening

updating to  "pinia": "2.2.0", causes the type error.

![grafik](https://github.com/user-attachments/assets/54d860ab-8589-4ad0-a556-b7d855aedc3a)

_*There's been no changes outside of the terminal screenshota other than the different pin of pinia in package.json_

It should also be noted that the same function (`something()`) does work fine in the script context.
![grafik](https://github.com/user-attachments/assets/b9d116d4-b529-4d99-80cf-3104228577dc)

**Reply by J-Michalek** [reply]
We are using both unplugin-auto-import and unplugin-vue-components, but I was only suspicious about unplugin-vue-components because I only saw errors that mentioned components.

**Reply by xmatthias** [reply]
it mentions components - but what fails is the auto-imported utils (uppon further, deeper inspection).
my initial thought was "believe the issue" until i thought about it further (in the car to work) 😆

**Reply by posva** [reply]
Thanks for the repro. I looked, and the error appears only with `vueTemplate: true`, and it happens no matter the `declare module` version used. I think it's worth trimming down the issue further to report it to vue-tsc or the auto import plugin

---

FYI about issues:

![Screenshot 2024-07-31 at 14 09 36](https://github.com/user-attachments/assets/f8668dad-3656-4a1a-8fd9-49a67543b15a)

**Reply by mxp131011** [reply]
> Thanks for the repro. I looked, and the error appears only with , and it happens no matter the version used. I think it's worth trimming down the issue further to report it to vue-tsc or the auto import plugin`vueTemplate: true``declare module`
> 
> FYI about issues:
> 
> ![Screenshot 2024-07-31 at 14 09 36](https://private-user-images.githubusercontent.com/664177/353824203-f8668dad-3656-4a1a-8fd9-49a67543b15a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0MzA3NzcsIm5iZiI6MTcyMjQzMDQ3NywicGF0aCI6Ii82NjQxNzcvMzUzODI0MjAzLWY4NjY4ZGFkLTM2NTYtNGExYS04ZmQ5LTQ5YTY3NTQzYjE1YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDczMVQxMjU0MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NDhjYTAwODFiYjNjN2QzZTE4NTYxZjRmNGYwZTk0NWE3ZmNkYzc5ZGNhNDA4ODc0MDVkMWE4Njg1MWNjYjI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ji1KKM0NHUPEpR5r-Xbbl0J35NOR8B0Uiy1YQZVNI-4)

我认为只有 2.2.0才会出现，在上一个版本我同样开启`vueTemplate: true`，它是正常的，只有在2.2.0 不正常，原因似乎于https://github.com/vuejs/language-tools/issues/4501  有关，解决方式就是

```json

"vueCompilerOptions": {
    "lib": "@vue/runtime-core",
},

```

**Reply by mxp131011** [reply]
In the previous version I also enabled vueTemplate: true, it is normal, only in 2.2.0 it is not normal, the reason seems to be related to [vuejs/language-tools#4501](https://github.com/vuejs/language-tools/issues/4501), the solution is

```json


"vueCompilerOptions": {
    "lib": "@vue/runtime-core",
},


```

**Reply by xmatthias** [reply]
@mxp131011 isn't this just inverting behavior from the previous version - which declared types for `@vue/runtime-core`

I doubt that workaround will work for both 2.2.0 and 2.1.7 - as it seems to simply invalidate the change that was made in 2.2.0 - which i assume was made to fix "something" (though not sure what).

**Reply by xmatthias** [reply]
>  the error appears only with vueTemplate: true, and it happens no matter the declare module version used

@posva with "declare module version used" you mean the change that was made in https://github.com/vuejs/pinia/commit/8a6ce86db83b6315c067c8a98c898b3c74efe62e#diff-094806c433011456914201ea662b339b53b9b1236f3226c1da3fd807106ad8a8R43 ?

**Reply by posva** [reply]
Yes

**Reply by xmatthias** [reply]
so you're implying that it's caused by something else - which would be curious - as then it's really a pinia bug - as pinia is the only thing changing.

**Reply by posva** [reply]
>  as then it's really a pinia bug - as pinia is the only thing changing.

I wished the web and Vue ecosystem was that simple!

FYI, the change of the declare module **was necessary in pinia's codebase, and it's also the recommended approach**, so it's not something I can _just_ revert and it's also why I'm suspecting the issue to be somewhere else. If you want to help, digging further to see why this error appears and opening a PR to the relevant repository would really move things forward!

**Reply by xmatthias** [reply]
I'm not sure i can replicate what you're saying above.

using pinia 2.2.0 and hot-patching the pinia module - i can get "no type-check" back again

![image](https://github.com/user-attachments/assets/90a2d05e-d74c-40e9-819b-049777429330)


---

have you got a doc reference (would be vue docs?) to where this approach is recommended so i can dig deeper? i honestly did struggle to find a proper reference to shi.

**Reply by posva** [reply, marked as answer]
BTW: updating vue-router's d.ts to also use `declare module 'vue'` makes everything work. Using multiple `declare module` at the same time like auto imports breaks the in template auto imports.

Edit: released vue-router with the update and it's fixed

**Reply by posva** [reply]
The docs about augmenting: https://vuejs.org/guide/typescript/options-api.html#augmenting-global-properties

**Reply by xmatthias** [reply]
Quite interesting - auto-import seems to export **both** - `declare module 'vue' {` as well  as `declare module '@vue/runtime-core' {` in their auto-generated file ... most likely to avoid some odd behavior already.

The update in vue-router does fix it though :+1:

**Reply by J-Michalek** [reply]
Thank you guys for resolving it!

**Reply by cooper667** [reply]
This fix _breaks_ one of our repos (but fixes another 🤷 ) with effectively the same dependencies (as far as i can tell, the ones that seem relevant to me match exactly. I'm basically just commenting here to see if this has affected anyone else!

(Commenting out `export interface GlobalComponents {}`, tsc is happy, with it in, there's all kinds of chaos, mostly around generic vue components it seems)

All of this seems like a higher vue or TS augmenting issue really

**Reply by xmatthias** [reply]
the latest nuxt release log mentions this change [explicitly](https://nuxt.com/blog/v3-13#vue-typescript-changes).

More importantly though - it suggests a (most likely non-nuxt exclusive) workaround for this in creating a `declarations.dts` file at the root of the project with the following content

```
import type {
  ComponentCustomOptions as _ComponentCustomOptions,
  ComponentCustomProperties as _ComponentCustomProperties,
} from 'vue';

declare module '@vue/runtime-core' {
  interface ComponentCustomProperties extends _ComponentCustomProperties {}
  interface ComponentCustomOptions extends _ComponentCustomOptions {}
}
```

Now i did neither test this (i've worked around it in other ways / been fortunate to have libraries update in time) - but considering it's mentioned in the nuxt changelog - I'd assume that it's working.

![image](https://github.com/user-attachments/assets/ca3896f5-13b0-4a34-84c5-55a50ebf3b2a)
